### ISSUE TITLE
Simplify the Data API

### ISSUE THREAD

**Issue by fsimonis** [precice/precice#1559]
**Please describe the problem you are trying to solve.**

The API to access data is cumbersome as it implements all possibilities of the cross-product  
`{read, write, writeGradient} x {Single, Block} x {Scalar, Vector}`

This leads to 12 API calls that are cumbersome to maintain as the input checks need to preserve context while their implementation logic is extremely similar.
It also makes using the API cumbersome as these functions slightly differ in syntax and return types.

**Describe the solution you propose.**

We use the block variant as a replacement for singular versions.
We omit the data type (Vector or Scalar) in the API name and deduce which of the two to use inside the function call. In the code, the scalar version can often be considered a vector of 1 element.
Currently, we check if the user calls the correct variant for the configured datatype. We can reuse this logic to select the correct process internally.

This would replace 12 API functions with:
```cpp
void SolverInterface::readData(         DataID did, int size, const int* indices, double* values);
void SolverInterface::writeData(        DataID did, int size, const int* indices, const double* values);
void SolverInterface::writeGradientData(DataID did, int size, const int* indices, const double* gradientValues);
```

Further upside is that this would simplify the documentation, as the code now doesn't explicitly refer to vectorial data any longer.

**Downside** for read-data is that the user can no-longer directly use/assign single values, but always need to use references using the more verbose variant.
```cpp
double d;
interface.readData(did, 1, &vertexID, &d); 

// previously
double d = interface.readScalarData(did, vertexID); 
```

An alternative could be a singular overload for the readScalarData version, which may result in confusion:
```cpp
void SolverInterface::readData(DataID did, int size, const int* indices, double* values);
double SolverInterface::readData(DataID did, int index);  // assumes scalar
```

**Describe alternatives you've considered**

Keep the Scalar/Vector differentiation, but merge Single/Block variants.
This at least reduced forwarding or other duplications.

**Comment by IshaanDesai** [comment]
As far as the Python bindings go, this simplification should not cause any issues. Actually it is more *pythonic* to deduce the data type (vector or scalar) and to not have the single and block variants but just one general API call. We already do this to some degree [in the fenics-adapter](https://github.com/precice/fenics-adapter/blob/9aa3e22496117725ba993ac677a1239dc6f156f8/fenicsprecice/fenicsprecice.py#L193-L230), wherein we deduce all the necessary information about the data within the `read_data` function.

**Comment by uekerman** [comment]
Another reason for such kind of simplification: There was some confusion of users expecting that `read_block_data` is some sort of blocking read, meaning it waits till new data arrives.

**Comment by uekerman** [comment]
While touching and breaking this, we should also check how consistent our naming schemes in the C and Fortran bindings is. 

For example:

https://github.com/precice/precice/blob/95a292c6e6f586d074b68bbde4b56394b49fc4bd/extras/bindings/fortran/src/SolverInterfaceFortran.cpp#L300

**Comment by BenjaminRodenberg** [comment]
> * move the waveform overloads of `readData` into a separate function `readDataAt` (name debatable).
>       **Reason:** there is no obvious sentinel value for `dt`. While in C++ we could use an `std::optional<double> dt` to model the semantics as we use in python (dt = None), this doesn't work in C/Fortran.

If the lack of a sentinel value in C/Fortran is the main reason: you could use a negative number (`dt=-1`), because using a negative `dt` is not allowed.

From my perspective there are two usability arguments for keeping the overload in the API: 

Currently the nice thing about overloading `readData` is that waveform relaxation is easy to turn on and off by just providing an optional argument. If we create a dedicated function for waveform relaxation I think it might become more difficult for users to switch from one mode to the other.

Also important: On the long run it is reasonable to always (at least implicitly) use some kind of waveform relaxation: Doing linear interpolation is for free and always possible. Additional runtime cost of waveform relaxation is an open question, but besides that there are no real drawbacks.

**Comment by uekerman** [comment]
Concerning waveform API: We could also go one step further and break backwards compatibility completely, meaning `readData` always requires an argument `dt`. I think this is where we want to get in the long run anyway. Switching WR then on/off should happen in the configuration.

**Comment by fsimonis** [comment]
@uekerman we could use 0 as a default value, right?
Combined with waveform of order 0, this will behave the same as the current readXData calls without the dt parameter.

**Comment by BenjaminRodenberg** [comment]
Another perspective on the discussion about `readData` and `dt` (sorry that it is a bit lengthy, but I think it's useful to have this documented somewhere):

Be aware that `dt=0` and `dt=timeWindowRemainder` always point to the begining and end of the time window. This means that we get here a potential source for inconsistencies: In the first time window `dt=timeWindowRemainder` points to the same point in time like `dt=0` in the second time window. If we say that a value is constant on the whole time window, this leads to a different value at the window boundary depending on whether we are looking at it from the window before or after the boundary.

If we are talking about windows this situation somehow works (this is actually the preCICE v2 way of dealing with time: There is no time, we only have one value on each window). But if we look at subcycling with piecewise constant interpolation, we need a solution. This is why I decided in https://github.com/precice/precice/pull/1523/files# to exclude the beginning of each time step (or window) here. See [implementation here](https://github.com/BenjaminRodenberg/precice/blob/951d87f86789ec7e994a873fc84b291b7028b42d/src/time/Waveform.cpp#L65-L67) and a [corresponding test](https://github.com/BenjaminRodenberg/precice/blob/951d87f86789ec7e994a873fc84b291b7028b42d/src/time/tests/WaveformTest.cpp#L83-L85). If we want to arrive at a consistent API this means: If I do piecewise constant interpolation the area where the constant value *lives* should always exclude the beginning of the interval and include the end. Or expressed with a formula: 

$$f(t) = v_i\text{, if } t \in (t_{i-1}, t_i]$$

I think this also makes more sense, if you consider data initialization (basically providing the starting value) and linear interpolation (where $f(t_{i-1}) = v_{i-1}$ and $f(t_i) = v_i$ are the interpolation conditions). In https://github.com/precice/precice/pull/1422 I also changed the behavior of the interface and some tests correspondingly.

**Comment by uekerman** [comment]
> @uekerman we could use 0 as a default value, right?

Could be confusing. The backwards compatible variant would be to use `dt=timeWindowRemainder` as default (see @BenjaminRodenberg's comment).

My suggestion would be not to provide a default value, but make `dt` mandatory.

It could be a hard ask from users, but we need this eventually.

**Comment by BenjaminRodenberg** [comment]
As far as I understand the discussion about the default value for `dt` there is only a problem for C/Fortran. I find it a bit hard to take a purely technical reason for some languages as an argument for a change of the API and usability for the whole ecosystem.

Why not keep everything as it is in C++, python and all other languages that allow us to implement a default value and only make an exception for the C/Fortran bindings. If we make `dt` mandatory, this means that users (and we) have to modify every single read data call:

```python
...
precice_dt = precice.advance()
read_data = read_block_scalar_data("myMesh", "myData", ids, dt=precice_dt)
...
```

I think the API is already complicated enough here and if we can hide the `dt` argument, we should also do it. Otherwise this will just be another argument users always have to provide for no clear reason.

**Comment by fsimonis** [comment]
@BenjaminRodenberg That's why I initially suggested to use 2 separate function names.
This makes it clearer what is going on and prevents name conflicts.

Of course, we could just use the separate names in the C and Fortran bindings.

**Comment by BenjaminRodenberg** [comment]
How about using the `readDataAt` approach in C, Fortran and Rust and default values elsewhere? We have to be careful that the different language bindings are not diverging too much, but I think it makes sense to allow for some flexibility here looking at the different available features in different languages (spans in C++, numpy arrays in python and so on).

**Comment by BenjaminRodenberg** [comment]
Suggestion: Let's implement the mandatory `dt` approach independent from this issue. I generally like the idea to make `dt` mandatory, if we really want to force our users to think more about the whole topic of read time. But I also see this as a different issue. So I opened #1603 to continue over there.

**Comment by fsimonis** [comment]
Another alternative is to readDataAt, is to make the delta version the preferred case:

```cpp
std::size_t readData(string_view meshName, string_view dataName,
    span<const int> vertexIds, double dt span<double> data) const;

std::size_t readDataAtTimeWindowEnd(string_view meshName, string_view dataName,
    span<const int> vertexIds, span<double> data) const;
```

**Comment by davidscn** [comment]
Discussing in person:

Now that we concluded in #1603, we would be left with these three functions

```cpp
void readData(string_view meshName, string_view dataName,
    span<const int> vertexIds, span<double> data) const;

void writeData(string_view meshName, string_view dataName,
    span<const int> vertexIds, span<const double> data);

void writeGradientData(string_view meshName, string_view dataName,
    span<const int> vertexIds, span<const double> gradient);
```

Since we also decided in #1577 to use a `precice::span`, we can provide the same user-safety compared to the current preCICE version, if we are strict regarding the accepted `vertexIds.size()` and `data.size()`, i.e., we only accept consistent IDs and corresponding data buffers.

However, implementing this approach requires to add getter functions for the number of data components used in order to allow applying this in the `C` and `Fortran` bindings. The function could look as follows:

```cpp
int getDataDimensions( precice::string_view meshName, precice::string_view dataName );
```

and is trivial to add. The counterpart of this function will then become `getMeshDimensions(...)`. Hence, we will change the data API as described.
